akka-docs/src/test/scala/docs/stream/operators/sourceorflow/MapAsyncs.scala

```diff
@@ -0,0 +1,123 @@
+package docs.stream.operators.sourceorflow
+
+import java.util.concurrent.TimeUnit
+
+import akka.NotUsed
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.scaladsl.Behaviors
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.util.Timeout
+
+import scala.concurrent.Future
+import scala.concurrent.duration._
+
+/**
+ *
+ */
+object CommonMapAsync {
+  case class Event(sequenceNumber: Int)
+  case class EventProcessingRequest(evt: Event, replyTo: ActorRef[Int])
+
+  val guardian = Behaviors.receiveMessage[EventProcessingRequest] {
+    case EventProcessingRequest(evt, replyTo) =>
+      TimeUnit.MILLISECONDS.sleep(500)
+      replyTo.tell(evt.sequenceNumber)
+      Behaviors.same
+  }
+
+  implicit val sys =
+    ActorSystem.apply[EventProcessingRequest](guardian, "mapAsync-stream")
+  implicit val exCtx = sys.executionContext
+  implicit val timeout: Timeout = 3.seconds
+
+  // #mapasync-strict-order
+  // #mapasync-concurrent
+  // #mapasyncunordered
+
+  val events: Source[Event, NotUsed] = //...
+    // #mapasync-strict-order
+    // #mapasync-concurrent
+    // #mapasyncunordered
+    Source.fromIterator(() => Iterator.from(1)).throttle(1, 50.millis).map { in =>
+      Event(in)
+    }
+
+  // #mapasync-strict-order
+  // #mapasync-concurrent
+  // #mapasyncunordered
+
+  def eventHandler(event: Event): Future[Int] = {
+    println(s"Processing event $event...")
+    //...
+    // #mapasync-strict-order
+    // #mapasync-concurrent
+    // #mapasyncunordered
+    val result =
+      if (event.sequenceNumber % 5 == 0) {
+        import akka.actor.typed.scaladsl.AskPattern._
+        sys.ask { replyTo =>
+          EventProcessingRequest(event, replyTo)
```

Comment:
perhaps instead of having and actor and using 'ask', we could use akka.pattern.after (https://doc.akka.io/docs/akka/current/futures.html#after) to implement the `eventHandler`.

Perhaps we could even make it deterministic by starting with a `Source(10, 1, 5, 2)` and make the processing complete the future after 10/1/5/2 seconds respectively?

Reply:
Oh, I just realized I didn't commit the scala version where the delay happens randomly (this happens on every 5th element).

Let me sync both samples first and we can see if making the delays deterministic helps.